### ISSUE TITLE
[Feature Request] Emit events for all data mutations

### ISSUE THREAD

**Issue by hstove** [blockstack/stacks-blockchain#1690]
**Is your feature request related to a problem? Please describe.**
I've been thinking about https://github.com/blockstack/stacks-blockchain-sidecar/issues/141, and what would be required to enable this functionality. After digging more into the sidecar, it seems like part of the problem is a lack of events for operations like `map-set`, etc.

**Describe the solution you'd like**
Emit events for functions that update a contract's state. `map-set`, `var-set`, at the least.

**Describe alternatives you've considered**
From the sidecar issue:

> I know that I could hack around this by keeping an in-Clarity map like my-map-entries, where the key is (index uint) and the value is the key for that map entry. Then, I could use the RPC API to get each map key by index, and then query for the value of that map key. However, that's hugely inefficient and goes straight to the core node, both of which are not ideal.

**Additional context**
I think enabling this would allow the sidecar to become quite powerful as a DX tool. Many hypothetical dapps that I can think of require more nuanced and complex queries than are (or should be) provided by the node's RPC API.

Basically, enabling this functionality allows the sidecar to potentially materialize contract state in a much more flexible and queryable fashion.

**Comment by jcnelson** [comment]
I'm confused about the underlying problem in this issue.  Is the problem that you need a way to step through the execution of a Clarity program, i.e. to see the sequence of inserts and gets it does on a data map?  Because if so, this could be better handled by a debugger for Clarity, without needing to involve the node or side-car at all.

Or, per the referenced issue, is the lack of pagination over maps the underlying problem?  I don't think Clarity state is ever going to be paginated by the node in the foreseeable future -- it's too expensive to add an index for each and every data map that gets created, and Clarity can't benefit from it since it can't iterate over lists whose sizes aren't known at instantiation-time.  But I don't think emitting an event for each data mutation is the right answer here either -- by definition, any intermediate mutations are necessarily temporary, since they occur within an (atomic) transaction.  Maybe transaction receipts could contain a list of map key/values added and removed by the transaction, like a "patch" to the side-car's materialized view?

**Comment by diwakergupta** [comment]
@hstove @zone117x can this be closed (since blockstack/stacks-blockchain-api#141 was addressed)?

**Comment by hstove** [comment]
Good to close for me!
